File: container.cpp
#include "container.h"

// constructor
Container::Container() {
    container_length_ = 0;
}

// destructor
Container::~Container() {
    Clear();
}

// clear container
void Container::Clear() {
    for(int i = 0; i < container_length_; i++) {
        delete cont_[i];
    }
    container_length_ = 0;
}

// set color and type of shape
bool Container::SetColorAndType(int shape_key, int color_key) {
    switch(shape_key) {
        case 1:
            cont_[container_length_] = new Rectangle();
            cont_[container_length_]->SetColor(color_key);
            break;
        case 2:
            cont_[container_length_] = new Triangle();
            cont_[container_length_]->SetColor(color_key);
            break;
        case 3:
            cont_[container_length_] = new Circle();
            cont_[container_length_]->SetColor(color_key);
            break;
        case 4:
            return false;
    }
    return true;
}

// input data for container
void Container::Input(ifstream &ifst) {
    while(!ifst.eof()) {
        int shape_key, color_key;
        ifst >> shape_key >> color_key;
        if(SetColorAndType(shape_key, color_key)) {
            cont_[container_length_]->Input(ifst);
            container_length_++;
        }
    }
}

// get random data for container
void Container::InputRandom(int size) {
    while(container_length_ < size) {
        auto shape_key = rand() % 3 + 1;
        auto color_key = rand() % 7 + 1;
        SetColorAndType(shape_key, color_key);
        switch(shape_key) {
            case 1:
                shape_key = ShapeName::RECTANGLE;
                cont_[container_length_]->InputRandom();
            case 2:
                shape_key = ShapeName::TRIANGLE;
                cont_[container_length_]->InputRandom();
            case 3:
                shape_key = ShapeName::CIRCLE;
                cont_[container_length_]->InputRandom();
        }
        container_length_++;
    }
}

// output data of container
void Container::Output(ofstream &ofst) {
    ofst << "Container contains " << container_length_ << " elements." << '\n';
    for(int i = 0; i < container_length_; i++) {
        ofst << i << ": ";
        ofst << "It is ";
        cont_[i]->OutColor(ofst);
        cont_[i]->Output(ofst);
    }
}

// sort container's elements by their perimeter
void Container::SortContainer() {
    for(int i = 0; i < container_length_; i++) {
        for(int j = container_length_ - 1; j > i; j--) {
            if(cont_[j]->Perimeter() < cont_[j - 1]->Perimeter()) {
                std::swap(cont_[j], cont_[j - 1]);
            }
        }
    }
}

File: container.py
from rectangle import Rectangle
from triangle import Triangle
from circle import Circle
from helper import Helper


class Container:
    def __init__(self):
        self.figures = []

    def set_shape(self, data, shape_type, shape_color):
        if len(data) != Helper.get_input_amount_for_shape(shape_type):
            return False
        if shape_type == 1:
            self.figures.append(Rectangle(shape_color, data))
        elif shape_type == 2:
            self.figures.append(Triangle(shape_color, data))
        elif shape_type == 3:
            self.figures.append(Circle(shape_color, data))
        else:
            return False
        return True

    def output(self, file):
        self.sort_container()
        size = len(self.figures)
        file.write("Container contains " + str(size) + " elements \n")
        index = 0
        for figure in self.figures:
            file.write(str(index) + ": ")
            figure.output(file)
            index += 1

    def sort_container(self):
        size = len(self.figures)
        for i in range(0, size):
            for j in range(size - 1, i, -1):
                if self.figures[j].get_perimeter() < self.figures[j - 1].get_perimeter():
                    self.figures[j], self.figures[j - 1] = self.figures[j - 1], self.figures[j]
